cartography/intel/aws/kms.py

```diff
@@ -0,0 +1,333 @@
+    # policy JSON format. Note condition can be any JSON statement so will need to import as-is
+    # policy is a very complex format, so the policyuniverse library will be used for parsing out important data
+    # ...metadata...
+    # "Policy" :
+    # {
+    #   "Version": "2012-10-17",
+    #   "Id": "key-consolepolicy-5",
+    #   "Statement": [
+    #     {
+    #       "Sid": "Enable IAM User Permissions",
+    #       "Effect": "Allow",
+    #       "Principal": {
+    #         "AWS": "arn:aws:iam::123456789012:root"
+    #       },
+    #       "Action": "kms:*",
+    #       "Resource": "*"
+    #     },
+    #     {
+    #       "Sid": "Allow access for Key Administrators",
+    #       "Effect": "Allow",
+    #       "Principal": {
+    #         "AWS": "arn:aws:iam::123456789012:role/ec2-manager"
+    #       },
+    #       "Action": [
+    #         "kms:Create*",
+    #         "kms:Describe*",
+    #         "kms:Enable*",
+    #         "kms:List*",
+    #         "kms:Put*",
+    #         "kms:Update*",
+    #         "kms:Revoke*",
+    #         "kms:Disable*",
+    #         "kms:Get*",
+    #         "kms:Delete*",
+    #         "kms:ScheduleKeyDeletion",
+    #         "kms:CancelKeyDeletion"
+    #       ],
+    #       "Resource": "*"
+    #     }
+    #   ]
+    # }
+    if policy is not None:
+        # get just the policy element and convert to JSON because boto3 returns this as string
+        policy = Policy(json.loads(policy['Policy']))
+        if policy.is_internet_accessible():
+            return {
+                "kms_key": key,
+                "internet_accessible": True,
+                "accessible_actions": list(policy.internet_accessible_actions()),
+            }
+        else:
+            return None
+    else:
+        return None
+
+
+@timeit
+def load_kms_keys(neo4j_session, data, region, current_aws_account_id, aws_update_tag):
+    ingest_keys = """
+    MERGE (kmskey:KMSKey{id:{KeyId}})
+    ON CREATE SET kmskey.firstseen = timestamp(),
+    kmskey.arn = {Arn}, kmskey.creationdate = {CreationDate}
+    SET kmskey.deletiondate = {DeletionDate},
+    kmskey.validto = {ValidTo},
+    kmskey.enabled = {Enabled},
+    kmskey.keystate = {KeyState},
+    kmskey.customkeystoreid = {CustomKeyStoreId},
+    kmskey.cloudhsmclusterid = {CloudHsmClusterId},
+    kmskey.lastupdated = {aws_update_tag},
+    kmskey.region = {Region}
+    WITH kmskey
+    MATCH (aa:AWSAccount{id: {AWS_ACCOUNT_ID}})
+    MERGE (aa)-[r:RESOURCE]->(kmskey)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {aws_update_tag}
+    """
+
+    # The owner data returned by the API maps to the aws account nickname and not the IAM user
+    # there doesn't seem to be a way to retreive the mapping but we can get the current context account
+    # so we map to that directly
+
+    for kms_key in data:
+        neo4j_session.run(
+            ingest_keys,
+            KeyId=kms_key["KeyId"],
```

Comment:
What is the difference between a KeyId and an Arn?

For AWS syncs, we should to do something like 

```
ON CREATE SET kmskey.id = kms_key.arn,
kmskey.firstseen = timestamp()
SET kmskey.arn = kms_key.arn,
[...]
```

It's a bit duplicative, but at least viewers of your data will be able to tell what field(s) the `id` was based off of.

Reply:
KeyId is just the Key Identifier, whereas ARN is fully formed URI for the Key. While populating Neo4j, we are using KeyId as the primary identifier.

As suggested, we have used UNWIND method for populating keys, grants and aliases.